Record metrics describing time spent queued

carterkozak opened the pull request:
This fixes a bug where tracing spans were not recorded when queued
requests were canceled.

## Before this PR
Only tracing provided insight into the time requests spent queued.

## After this PR
==COMMIT_MSG==
Record metrics describing time spent queued
==COMMIT_MSG==

## Possible downsides?
more complexity surrounding the queue.

carterkozak committed:
Add generated changelog entries

carterkozak committed:
Record metrics describing time spent queued

This fixes a bug where tracing spans were not recorded when queued
requests were canceled.

carterkozak commented on dialogue-core/src/main/java/com/palantir/dialogue/core/QueuedChannel.java:
```diff
@@ -92,6 +96,8 @@
                 ListenableFuture<Response> result = maybeResult.get();
                 DialogueFutures.addDirectListener(result, this::onCompletion);
+                // While the queue was avoid, this is equivalent to spending zero time on the queue.
+                queuedTime.update(0, TimeUnit.NANOSECONDS);
```
I could go either way on this, but I opted toward updating with zero time to avoid biasing our statistics toward problematic regions. This code path is an optimization to avoid unnecessary work in the queue unless requests are already queued.

iamdanfox commented on dialogue-core/src/main/java/com/palantir/dialogue/core/QueuedChannel.java:
```diff
@@ -65,6 +67,7 @@
     private final AtomicInteger queueSizeEstimate = new AtomicInteger(0);
     private final int maxQueueSize;
     private Counter queueSizeCounter;
+    private Timer queuedTime;
```
why would this not be final?

carterkozak replied:
doh, this and the counter should be final

carterkozak committed:
final

carterkozak committed:
singular

iamdanfox reviewed (approved):
LGTM, can probably also amend the TODO:
```
TODO(jellis): record metrics for queue sizes, num requests in flight, time spent in queue, etc.
```

ferozco commented:
👍

carterkozak committed:
todone

carterkozak commented:
good call, I've removed the todo because we've implemented the relevant queue metrics, in-flight is out of scope for the queue and already recorded elsewhere.

svc-autorelease commented:
Released [0.19.11](https://github.com/palantir/dialogue/releases/tag/0.19.11)
